### PR TITLE
[Quant] Refactor reference module mapping

### PR DESCRIPTION
**Pull request by terrychenism**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #72755

Summary:
1. Add is_reference flag in convert function
if is_reference = true, using the default reference module mapping

2. add linear reference module and test case

Test Plan:
python3 test/test_quantization.py TestQuantizeEagerOps.test_conv_transpose_2d

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34188856](https://our.internmc.facebook.com/intern/diff/D34188856)

### CODE REVIEW

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34188856).

File: torch/ao/quantization/quantize.py [hunk 1]
```diff
@@ -500,6 +501,8 @@ def convert(
 
     """
     torch._C._log_api_usage_once("quantization_api.quantize.convert")
+    if is_reference:
+        mapping = get_default_static_quant_reference_module_mappings()
```
Comment (jerryzh168) [inline torch/ao/quantization/quantize.py:505]: I think we should do this in _convert, to keep everything in the same place

File: torch/nn/quantized/_reference/modules/linear.py [hunk 2]
```diff
@@ -15,6 +15,8 @@ class Linear(nn.Linear):
     and dequantize the weight before running the floating point functional
     linear operator.
     """
+    _IS_REFERENCE = True
```
Comment (jerryzh168) [inline torch/nn/quantized/_reference/modules/linear.py:18]: is linear not tested right now? should we add a test?

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34188856).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34188856).

File: torch/ao/quantization/quantize.py [hunk 3]
```diff
@@ -472,7 +473,7 @@ def quantize_qat(model, run_fn, run_args, inplace=False):
 
 def convert(
         module, mapping=None, inplace=False, remove_qconfig=True,
-        convert_custom_config_dict=None):
+        is_reference=False, convert_custom_config_dict=None):
     r"""Converts submodules in input module to a different module according to `mapping`
```
Comment (jerryzh168) [inline torch/ao/quantization/quantize.py:477]: maybe add some docs for `is_reference` flag?

**Review by jerryzh168 — APPROVED** [review]
looks good, would be better to have a more detailed summary message as well. e.g. talk about the motivation and impact of the change and how the is_reference flag is implemented

File: torch/ao/quantization/quantize.py [hunk 4]
```diff
@@ -524,6 +525,8 @@ def _convert(
                  is mutated
 
     """
+    if is_reference:
+        mapping = get_default_static_quant_reference_module_mappings()
```
Comment (jerryzh168) [inline torch/ao/quantization/quantize.py:529]: actually not like this, this should go within the L530 branch I think

Comment (jerryzh168) [reply torch/ao/quantization/quantize.py:529]: user should be able to overwrite the default mapping with explicit mappings, if user do not provide any mapping, we can provide a default mapping based on `is_reference` flag

**Review by jerryzh168 — CHANGES_REQUESTED** [review]
need to change the place for `is_reference` check in `_convert` function

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34188856).

File: torch/ao/quantization/quantize.py [hunk 5]
```diff
@@ -522,10 +523,13 @@ def _convert(
                  Modules
         inplace: carry out model transformations in-place, the original module
                  is mutated
+        is_reference: a flag to enable quantized reference module
 
     """
     if mapping is None:
         mapping = get_default_static_quant_module_mappings()
+        if is_reference:
```
Comment (jerryzh168) [inline torch/ao/quantization/quantize.py:531]: nit: the following is a little bit better
```
if is_reference:
   ...
else:
   ...
```

Comment (jerryzh168) [reply torch/ao/quantization/quantize.py:531]: or just a single line with
```
mapping = xxx if is_reference else yyy
```

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34188856).

**Review by jerryzh168 — APPROVED** [review]

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34188856).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34188856).

**Comment by terrychenism** [comment]
@terrychenism has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34188856).

**Pull request closed without merging**